feat(1) : 관리자 신청 휴가 조회 및 필터링 검색 서비스 구현

gunwoong1630 opened the pull request:
제목: feat(1): 관리자 휴가 조회 서비스 구현 - ApprovalStep 구현 및, 조회, 필터링 검색

## #️⃣연관된 이슈 번호
- #11 

## 📝작업 내용

erd cloud 내용을 토대로 VacationRequest와 ApprovalStep 구현하였음

``` java

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ApprovalStep extends BaseEntity {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "approval_step_id")
	private Long approvalStepId;

	private int step;
	@Column(name = "approval_status")
	@Enumerated(value = EnumType.STRING)
	private ApprovalStatus approvalStatus;

	@ManyToOne(cascade = CascadeType.PERSIST)
	@JoinColumn(name = "member_id")
	private Member member;

	@ManyToOne(cascade = CascadeType.PERSIST)
	@JoinColumn(name = "vacation_request_id")
	private VacationRequest vacationRequest;

	@Builder
	public ApprovalStep(int step, ApprovalStatus approvalStatus, Member member, VacationRequest vacationRequest) {
		this.step = step;
		this.approvalStatus = approvalStatus;
		this.member = member;
		this.vacationRequest = vacationRequest;
	}

	public void apply(ApprovalStatus approvalStatus) {
		this.approvalStatus = approvalStatus;
	}
}

public enum ApprovalStatus {
	IN_PROGRESS,
	DRAFT,
	APPROVED,
	REJECTED
}

```

``` java

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class VacationRequest extends BaseEntity {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "vacation_request_id")
	private Long id;

	@Column(name = "from_date", nullable = false)
	private LocalDate from;

	@Column(name = "to_date", nullable = false)
	private LocalDate to;

	private String reason;
	private int lastApprovalStep;

	@ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
	@JoinColumn(name = "type_code")
	private Code type;

	@Enumerated(value = EnumType.STRING)
	private VacationRequestStatus status;

	@ManyToOne(cascade = CascadeType.PERSIST)
	@JoinColumn(name = "member_id")
	private Member member;

	@Version
	private Integer version;

	@Builder
	public VacationRequest(LocalDate from, LocalDate to, String reason, Code type,
		Member member) {
		this.from = from;
		this.to = to;
		this.reason = reason;
		this.lastApprovalStep = 0;
		this.type = type;
		this.status = VacationRequestStatus.IN_PROGRESS;
		this.member = member;
	}

	public void updateStatus(VacationRequestStatus vacationRequestStatus) {
		this.status = vacationRequestStatus;
	}

	public void processVacationRequest(int lastApprovalStepIdx) {
		if (this.lastApprovalStep < lastApprovalStepIdx) {
			this.lastApprovalStep++;
		}
	}
}

public enum VacationRequestStatus {
	IN_PROGRESS,
	APPROVED,
	REJECTED
}

```


필터링을 위해 우선 필터링 대상과 관리자에게 보여줘야할 객체(DTO)를 아래와 같이 정의

``` java

public record AdminVacationSearchCondition(
	DateRangeCondition dateRange,
	ApplicantCondition applicant,
	ApproverCondition approver,
	// 휴가 신청 상태
	VacationRequestStatus vacationRequestStatus
) {

	public record DateRangeCondition(
		// 시작일
		LocalDate start,
		// 종료일
		LocalDate end,
		// 년도
		Integer year,
		// 분기
		Quarter quarter
	) {
	}

	// 휴가 신청자
	public record ApplicantCondition(
		// 이름
		String name,
		// 부서명
		String deptName,
		// 직책 codeId
		Long positionCodeId,
		// 휴가 종류 codeId
		Long vacationTypeCodeId
	) {
	}

	// 휴가 결재자
	public record ApproverCondition(
		// 이름
		String name,
		// 직책 codeId
		Long positionCodeId
	) {
	}
}

```

``` java

public record VacationRequestReadResponse(String type, LocalDate from, LocalDate to, String applicantName,
										  String approverName, String deptName, VacationRequestStatus status) {
}

```

추가적으로 분기도 있으면 좋을것같고 시간이 별로 걸리지 않을것이라 판단하여 분기별 검색을 위한 Quarter 객체 추가

``` java

public enum Quarter {
	Q1(1), Q2(2), Q3(3), Q4(4), H1(5), H2(6), NONE(0);

	private final int value;

	Quarter(int value) {
		this.value = value;
	}

	public static Optional<LocalDate> getStart(Integer year, Quarter quarter) {
		if (year == null) {
			return Optional.empty();
		}
		if (quarter == null) {
			quarter = Quarter.NONE;
		}

		return switch (quarter) {
			case Q1, Q2, Q3, Q4 ->
				Optional.of(LocalDate.of(year, 1, 1).with(IsoFields.QUARTER_OF_YEAR, quarter.value).with(
					TemporalAdjusters.firstDayOfMonth()));
			case H1, NONE -> Optional.of(LocalDate.of(year, Month.JANUARY, 1));
			case H2 -> Optional.of(LocalDate.of(year, Month.JULY, 1));
		};
	}

	public static Optional<LocalDate> getEnd(Integer year, Quarter quarter) {
		if (year == null) {
			return Optional.empty();
		}
		if (quarter == null) {
			quarter = Quarter.NONE;
		}

		return switch (quarter) {
			case Q1, Q2, Q3, Q4 ->
				Optional.of(LocalDate.of(year, 1, 1).with(IsoFields.QUARTER_OF_YEAR, quarter.value).with(
					TemporalAdjusters.lastDayOfMonth()).plusMonths(2));
			case H1 -> Optional.of(LocalDate.of(year, Month.JUNE, 30));
			case H2, NONE -> Optional.of(LocalDate.of(year, Month.DECEMBER, 31));
		};
	}
}

```

필터링을 하기 위해 다양한 방법(Query, QueryDSL)을 찾아봤지만 여러 요인들로 인해 criteria를 활용한 필터링 선택(갓PT)
코드별 자세한 설명은 주석 참조

``` java

@Repository
@RequiredArgsConstructor
public class VacationRequestSearchCustom {
	private final VacationRequestRepository vacationRequestRepository;
	private final EntityManager entityManager;

	/**
	 * ApprovalStep와 VacationRequest를 join하고 AdminVacationSearchCondition의 변수들을 통해 다중 필터 구현
	 * From<A,B> = 'from 엔티티'에서 엔티티 부분 , A타입의 객체부터 B타입의 속성을 탐색 (그래서 Root<A,A>이고 Join<A,B>임, Root와 Join 둘다 From 상속)
	 * SingularAttribute<C,D> = 메타모델에서 사용하는 단일 속성 타입 정보 , C = 특정 엔티티, D = C 엔티티의 필드 타입
	 * Path<X> = 메타모델 경로 혹은 루트로부터 탐색된 속성(특정 속성의 경로) , X = 속성 타입, 해당 path가 가르키는 최종 필드 타입
	 * @param searchCondition
	 * @param pageable
	 * @return
	 */
	public Page<VacationRequestReadResponse> search(AdminVacationSearchCondition searchCondition, Pageable pageable) {
		CriteriaBuilder cb = entityManager.getCriteriaBuilder();
		CriteriaQuery<VacationRequestReadResponse> cq = cb.createQuery(VacationRequestReadResponse.class);

		Root<ApprovalStep> as = cq.from(ApprovalStep.class);
		Join<ApprovalStep, VacationRequest> vr = as.join("vacationRequest", JoinType.INNER);

		/**
		 * 1. 각 휴가 신청서마다 최종 결제자 기준으로 조회
		 * 2. 휴가 신청 범위
		 * 3. 특정 년도 혹은 특정 분기 (1,2,3,4,상,하반기)
		 * 4. 휴가 신청자 이름
		 * 5. 부서 이름
		 * 6. 휴가 종류
		 * 7. 휴가 신청자 포지션
		 * 8. 결재자 포지션
		 * 9. 휴가 신청 상태
		 * 10. 휴가 결재자이름
		 */
		List<Predicate> predicates = CriteriaCustomPredicateBuilder.<ApprovalStep>builder(cb)
			.applyEqualFilter(as, ApprovalStep_.step, vr, VacationRequest_.lastApprovalStep)
			.applyDateRangeFilter(vr, VacationRequest_.from, VacationRequest_.to,
				searchCondition.dateRange().start(), searchCondition.dateRange().end())
			.applyDateRangeFilter(vr, VacationRequest_.from, VacationRequest_.to, Quarter.getStart(
				searchCondition.dateRange().year(), searchCondition.dateRange().quarter()), Quarter.getEnd(
				searchCondition.dateRange().year(), searchCondition.dateRange().quarter()))
			.applyLikeFilter(vr, searchCondition.applicant().name(), VacationRequest_.member, Member_.name)
			.applyLikeFilter(vr, searchCondition.applicant().deptName(), VacationRequest_.member, Member_.dept,
				Dept_.deptName)
			.applyEqualFilter(vr, searchCondition.applicant().vacationTypeCodeId(), VacationRequest_.type, Code_.id)
			.applyEqualFilter(vr, searchCondition.applicant().positionCodeId(), VacationRequest_.member,
				Member_.position, Code_.id)
			.applyEqualFilter(as, searchCondition.approver().positionCodeId(), VacationRequest_.member,
				Member_.position, Code_.id)
			.applyEqualFilter(vr, searchCondition.vacationRequestStatus(), VacationRequest_.status)
			.applyLikeFilter(as, searchCondition.approver().name(), ApprovalStep_.member, Member_.name)
			.build();

		/**
		 * Predicates들을 기반을 Query 생성
		 */
		TypedQuery<VacationRequestReadResponse> query = CriteriaCustomQueryBuilder.builder(
				cq, cb)
			.applyDynamicPredicates(predicates)
			.projection(VacationRequestReadResponse.class, vr.get(VacationRequest_.type),
				vr.get(VacationRequest_.from),
				vr.get(VacationRequest_.to),
				vr.get(VacationRequest_.member).get(Member_.name),
				as.get(ApprovalStep_.member).get(Member_.name),
				vr.get(VacationRequest_.member).get(Member_.dept).get(Dept_.deptName),
				vr.get(VacationRequest_.status))
			.createQuery(entityManager)
			.build();

		return QueryUtils.makeQueryToPageable(query, pageable,
			vacationRequestRepository.count());
	}
}

```

``` java
/**
 * 필터링을 위한 Predicate 리스트 빌드
 * @author gunwoong
 */
public class CriteriaCustomPredicateBuilder<T> {
	private final CriteriaBuilder cb;
	private final List<Predicate> predicates;

	private CriteriaCustomPredicateBuilder(CriteriaBuilder cb) {
		this.cb = cb;
		this.predicates = new ArrayList<>();
	}

	public static <T> CriteriaCustomPredicateBuilder<T> builder(CriteriaBuilder cb) {
		return new CriteriaCustomPredicateBuilder<>(cb);
	}

	public <R> CriteriaCustomPredicateBuilder<T> applyDateRangeFilter(From<T, R> root,
		SingularAttribute<? super R, LocalDate> mappedFromField,
		SingularAttribute<? super R, LocalDate> mappedToField,
		LocalDate from, LocalDate to) {
		if (isProvided(from) && isProvided(to)) {
			this.predicates.add(cb.greaterThanOrEqualTo(root.get(mappedToField), from));
			this.predicates.add(cb.lessThanOrEqualTo(root.get(mappedFromField), to));
		}
		return this;
	}

	public <R> CriteriaCustomPredicateBuilder<T> applyDateRangeFilter(From<T, R> root,
		SingularAttribute<? super R, LocalDate> mappedFromField,
		SingularAttribute<? super R, LocalDate> mappedToField,
		Optional<LocalDate> from, Optional<LocalDate> to) {
		return applyDateRangeFilter(root, mappedFromField, mappedToField, from.orElse(null), to.orElse(null));
	}

	public CriteriaCustomPredicateBuilder<T> applyEqualFilter(From<T, ?> root, Object conditionValue,
		SingularAttribute<?, ?>... mappedFields) {
		if (isProvided(conditionValue)) {
			this.predicates.add(cb.equal(CriteriaUtils.searchPath(root, mappedFields), conditionValue));
		}
		return this;
	}

	public <X, A, B> CriteriaCustomPredicateBuilder<T> applyEqualFilter(From<X, A> root,
		SingularAttribute<? super A, ?> mappedField,
		From<X, B> anotherRoot, SingularAttribute<? super B, ?> mappedAnotherField) {
		this.predicates.add(cb.equal(root.get(mappedField), anotherRoot.get(mappedAnotherField)));
		return this;
	}

	public CriteriaCustomPredicateBuilder<T> applyLikeFilter(From<T, ?> root, String conditionValue,
		SingularAttribute<?, ?>... mappedFields) {
		if (isProvided(conditionValue)) {
			this.predicates.add(cb.like(CriteriaUtils.searchPath(root, mappedFields), "%" + conditionValue + "%"));
		}
		return this;
	}

	public List<Predicate> build() {
		return predicates;
	}

	private boolean isProvided(Object value) {
		return value != null;
	}
}
```

``` java
/**
 * 주어진 필터(Predicates)를 기반으로 쿼리 생성 및 projection
 * @param <T>
 * @author gunwoong
 */
public class CriteriaCustomQueryBuilder<T> {
	private CriteriaQuery<T> cq;
	private CriteriaBuilder cb;
	private TypedQuery<T> typedQuery;

	private CriteriaCustomQueryBuilder(CriteriaQuery<T> cq, CriteriaBuilder cb) {
		this.cq = cq;
		this.cb = cb;
	}

	public static <T> CriteriaCustomQueryBuilder<T> builder(CriteriaQuery<T> cq, CriteriaBuilder cb) {
		return new CriteriaCustomQueryBuilder<>(cq, cb);
	}

	public CriteriaCustomQueryBuilder<T> applyDynamicPredicates(List<Predicate> predicates) {
		cq.where(cb.and(predicates.toArray(new Predicate[0])));
		return this;
	}

	public CriteriaCustomQueryBuilder<T> projection(Class<T> projectionClass, Expression... projectionFields) {
		cq.select(cb.construct(projectionClass, projectionFields));
		return this;
	}

	public CriteriaCustomQueryBuilder<T> createQuery(EntityManager entityManager) {
		this.typedQuery = entityManager.createQuery(cq);
		return this;
	}

	public TypedQuery<T> build() {
		return typedQuery;
	}

}

```

``` java
/**
 * @author gunwoong
 */
@UtilityClass
public class CriteriaUtils {
	/**
	 * 연관관계가 있는 클래스들을 Root에서 연달아 찾아가야할 경우 위와같이 Path를 찾아냄
	 * @param <X>
	 */
	public static<X> Path<String> searchPath(From<X,?> root, SingularAttribute... mappedFields) {
		Path<String> path = root.get(mappedFields[0]);
		for (int i = 1; i < mappedFields.length; i++) {
			path = path.get(mappedFields[i]);
		}
		return path;
	}
}

/**
 * @author gunwoong
 */
@UtilityClass
public class QueryUtils {
	/**
	 * 제공된 쿼리를 페이징하여 페이지 객체 리턴
	 */
	public static <T> Page<T> makeQueryToPageable(TypedQuery<T> query, Pageable pageable, Long totalCount) {
		return new PageImpl<>(query.setFirstResult((int)pageable.getOffset())
			.setMaxResults(pageable.getPageSize())
			.getResultList(), pageable, totalCount);
	}
}
```
## 🧪 테스트 여부
- [x] 테스트 코드를 작성함
- [x] 테스트를 수행함
테스트 코드는 VacationRequestRepositorySearchTest 를 참조 바람

## 💬 참고사항
- 기능 구현 과정에서 해보지 않았던 검색 필터링의 기능 수행과정에서 테스트는 필수였음
- 관리자 기준 검색이기 때문에 전체적인 데이터를 조회하는 과정에서 일부 기존의 객체의 수정이 불가피 하였고 하필이면 commit이 섞였음
- 그럼으로 conflict가 예상되지만 기존 엔티티의 몇몇 변수에 nullable과 cascade만 수정했기 때문에 간단하게 conflict 해결할 것이라 예측

gunwoong1630 committed:
feat: ApprovalStatus 객체 추가 및 VacationRequest 필드 수정

1. ApprovalStatus 엔티티화
2. VacationRequest
- from과 to의 LocalDateTime => LocalDate (기간 비교 단순화를 위해)
- ApprovalStatus와 결제 정보 동기화를 위해 VacationRequest에 2개의 함수 추

gunwoong1630 committed:
refactor: 이후 테스트를 위해 cascade 추가 및 repository 부재 추가

gunwoong1630 committed:
feat: 관리자 측면 휴가 신청 조회 및 필터링 검색

Repository에서 Query로 진행하는 과정에서 복잡한 구조에 비해 확장성이 떨어져서 Repository를 따로 생성하고 Criteria를 활용하여 필터링 구현

필터링 내역
- 휴가 신청 범위
- 특정 분기 (1,2,3,4,상,하반기)
- 휴가 신청자 이름
- 부서 이름
- code id (휴가 종류, 휴가 신청자 포지션, 결재자 포지션)
- 휴가 신청 상태
- 휴가 결재자 이름

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/controller/VacationRequestController.java:
```diff
@@ -0,0 +1,30 @@
+import programmers.team6.domain.vacation.service.VacationRequestService;
+
+@RestController
+@RequestMapping("/api/v1/admin/vacationRequest")
```
잘 만들어주셨는데 버전닝 빼기로 했는데 들어가 있는 것 같아요 ㅠㅠ!

gunwoong1630 replied:
수정 완료

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/controller/VacationRequestController.java:
```diff
@@ -0,0 +1,30 @@
+public class VacationRequestController {
+	private final VacationRequestService vacationRequestService;
+
+	@PostMapping("/search")
```
조회로 알고있는데~ PostMapping을 사용하신 이유가 있으신가요!?
Rest명시상 GET에 가까워보여서 말씀드립니다!

gunwoong1630 replied:
수정완료

min0962 reviewed (approved):
필터링.. 저도 많이 찾아보다가 그냥 jpql 쌩조회로 어찌저찌 구현했는데,,
추후에 참고 해봐야겠네요..!

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/AdminVacationSearchCondition.java:
```diff
@@ -0,0 +1,35 @@
+
+@Getter
+@Setter
+@AllArgsConstructor
```
잘만들어주셨는데 혹시 Setter넣은신 이유가 있을까요?
약간 Dto에서 setter가 있으면 사용자 정보를 변경하는 느낌이 들어서! 빼면 어떨까 싶습니다!

gunwoong1630 replied:
수정완료

sehee123 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/ApprovalStatus.java:
```diff
@@ -0,0 +1,8 @@
+package programmers.team6.domain.vacation.dto;
```
이부분은 [enums]패키지를 따로 만들어서 관리하면 좋을 것 같습니다! 🙌

gunwoong1630 replied:
수정 완료

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/Quarter.java:
```diff
@@ -0,0 +1,40 @@
+		this.endDay = endDay;
+	}
+
+	public static Optional<LocalDate> getStartDate(Integer year,Quarter quarter) {
+		if (quarter == null || year == null) {
+			return Optional.empty();
+		}
+		return Optional.of(LocalDate.of(year, quarter.startMonth, quarter.startDay));
+	}
+
+	public static Optional<LocalDate> getEndDate(Integer year,Quarter quarter) {
+		if (quarter == null || year == null) {
+			return Optional.empty();
+		}
+		return Optional.of(LocalDate.of(year, quarter.endMonth, quarter.endDay));
+	}
```
static으로 작성하신 이유가 있는지 궁금합니다!  개인적으로
```java
	public Optional<LocalDate> getStartDate(Integer year) {
		if (year == null) {
			return Optional.empty();
		}
		return Optional.of(LocalDate.of(year,this.startMonth,this.startDay));
	}
```
위 처럼 작성하면 문제가 될까요?

gunwoong1630 replied:
이전에 위와같이 작성했는데, Quarter 자체가 null일 경우 문제가 생김으로 static으로 구현했습니다. 

만약 위와 같은 방식으로 한다면 Quarter쓰기전에 null check하고 Quarter에 들어와서도 null check해야되기 때문에, 한번에 묶어서 null check하고자 위와같은 방식으로 구현했습니다.

이후에 멘토님께 피드백받은 부분을 참고하여 바꿀 예정입니다

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/service/VacationRequestService.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface VacationRequestService {
+	Page<VacationRequestReadResponse> search(Pageable pageable,AdminVacationSearchCondition searchCondition);
+}
```
혹시 따로 Service를 인터페이스로 분리해서 사용하시는 이유가 있을까요!?

gunwoong1630 replied:
수정 완료

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/VacationRequestSearchCustom.java:
```diff
@@ -0,0 +1,125 @@
+	private final VacationRequestRepository vacationRequestRepository;
+	private final EntityManager entityManager;
+	private CriteriaBuilder cb;
+	private List<Predicate> predicates;
```
크라이테이라 생각도 못했는데 👍

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/VacationRequestSearchCustom.java:
```diff
@@ -0,0 +1,125 @@
+		return new PageImpl<>(result, pageable, totalCount);
+	}
+
+	private void addPredicateBetweenDates(From root, Optional<LocalDate> start,
+		Optional<LocalDate> end) {
+		if (start.isPresent() && end.isPresent()) {
+			predicates.add(cb.lessThanOrEqualTo(root.get("from"), end.get()));
+			predicates.add(cb.greaterThanOrEqualTo(root.get("to"), start.get()));
+		}
+	}
```
잘 작성해주셨는데 Optional을 인자로 넘기면 원래는 Optional에 null이 들어오는지도 체크해야하거든요 그래서 여기서는 Optional를 제거하는건 어떤가요?

gunwoong1630 commented:
지금 보니 조회할때 최신 시간순서대로 가져오는 기능을 깜빡하고 안넣었네요 빠른 시일 내로 추가하겠습니다

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/VacationRequestSearchCustom.java:
```diff
@@ -0,0 +1,125 @@
+		Join<ApprovalStep, VacationRequest> vr = as.join("vacationRequest", JoinType.INNER);
+		this.predicates = new ArrayList<>();
+		this.predicates.add(cb.equal(as.get("step"), vr.get("lastApprovalStep")));
+
+		// 휴가 신청 범위
+		addPredicateBetweenDates(vr, Optional.ofNullable(searchCondition.getStart()),
+			Optional.ofNullable(searchCondition.getEnd()));
+		// 특정 분기 (1,2,3,4,상,하반기)
+		addPredicateBetweenDates(vr, Quarter.getStartDate(searchCondition.getYear(), searchCondition.getQuarter()),
+			Quarter.getEndDate(searchCondition.getYear(), searchCondition.getQuarter()));
+		// 휴가 신청자 이름
+		addPredicateLike(vr, new String[] {"member", "name"}, searchCondition.getApplicantName());
+		// 부서 이름
+		addPredicateLike(vr, new String[]{"member", "dept","deptName"}, searchCondition.getDeptName());
+		// code id를 활용
+		if (searchCondition.getCodeId() != null) {
+			// 휴가 종류
+			Predicate vacationRequestType = cb.equal(vr.get("type").get("id"), searchCondition.getCodeId());
+			// 휴가 신청자 포지션
+			Predicate vacationRequestMemberPosition = cb.equal(vr.get("member").get("position").get("id"), searchCondition.getCodeId());
+			// 결재자 포지션
+			Predicate approvalStepMemberPosition = cb.equal(as.get("member").get("position").get("id"), searchCondition.getCodeId());
+			predicates.add(cb.or(vacationRequestType, vacationRequestMemberPosition, approvalStepMemberPosition));
+		}
+		// 휴가 신청 상태
+		addPredicateEqual(vr, new String[] {"status"}, searchCondition.getVacationRequestStatus());
+		// 휴가 결재자 이름
+		addPredicateLike(as, new String[] {"member", "name"}, searchCondition.getApproverName());
+
+		cq.where(cb.and(predicates.toArray(new Predicate[0])));
+
+		// projection
+		cq.select(cb.construct(VacationRequestReadResponse.class,
+			vr.get("type"),
+			vr.get("from"),
+			vr.get("to"),
+			vr.get("member").get("name"),
+			as.get("member").get("name"),
+			vr.get("member").get("dept").get("deptName"),
+			vr.get("status")
```
이건 필수는 아닌데 한번 생각해보셨으면 좋겠어요!
지금 크라이테리아 잘 찾아와서 써주셨는데 뭔가 상당히 크라이테리아 쪽 코드에 의존적이라서 읽기가 조금 힘든느낌이 있는데 CriterQueryBuilder같은걸 만드는건 어떨까요!?

gunwoong1630 replied:
수정 완료

rhwlgns4386 commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/VacationRequestSearchCustom.java:
```diff
@@ -0,0 +1,125 @@
+	private final VacationRequestRepository vacationRequestRepository;
+	private final EntityManager entityManager;
+	private CriteriaBuilder cb;
+	private List<Predicate> predicates;
```
predicates 공유해서 사용하고 있는데 이거 빈이라 싱글톤이다 보니 predicates가 멀티스레드에서 공유 되어 버립니다! 이부분을 로컬로 빼시는게 좋을 것 같아요!

rhwlgns4386 replied:
CriteriaBuilder도 마찬가지일것 같아요!

gunwoong1630 replied:
수정 완료

rhwlgns4386 reviewed (commented):
고생 너무 많은셨습니다! 리뷰보시고 수정사항 한번 생각해주시면 좋을 것 같아요!😄

Account-4-Review commented on backend/team6/src/main/java/programmers/team6/domain/vacation/dto/AdminVacationSearchCondition.java:
```diff
@@ -0,0 +1,35 @@
+@AllArgsConstructor
+public class AdminVacationSearchCondition {
+	// 휴가 신청 범위
+	private LocalDate start;
```
`@NotNull LocalDate start` <- notNull 일 때
이외의 경우엔 그냥 Record나 일반 dto 사용해도 문제 없습니다. (해당 필드가 nullable 인게 record와 연관 없음)

Account-4-Review replied:
jakarta validation 적용해서, 기본적인 값의 필터링을 적용하면 좋겠습니다.
(ex. `@NotNull`, `@NotEmpty`, `@Email`, `@Future`)

Account-4-Review commented on backend/team6/src/main/java/programmers/team6/domain/vacation/controller/VacationRequestController.java:
```diff
@@ -0,0 +1,30 @@
+	private final VacationRequestService vacationRequestService;
+
+	@PostMapping("/search")
+	Page<VacationRequestReadResponse> selectVacationRequests(
```
`@ResponseStatus(HttpStatus.OK)`

gunwoong1630 replied:
수정 완료

gunwoong1630 committed:
refactor: 코드 리뷰 기반 수정

1. AdminVacationSearchCondition setter 제거
2. enum 디렉토리 생성 및 이동, Quarter IsoField 활용하여 구현
3. criteria 수정
- 기존 방식(문자열 기반 검색)에서 메타모델 활용 방식으로 전환, 그리하여 문자열 기반의 방식의 런타임 과정에서 에러 발생 가능성을 낮춤
- hibernate-jpamodelgen gradle 추가하여 메타 모델을 자동 생성으로 전환
- criteria를 활용한 필터링 쿼리 코드 수정
    - 각 기능(필터링, 쿼리 생성, 쿼리 수행, 페이징)을 분리하여 필터링과 쿼리 생성은 빌더를 활용하여 리팩토링 수행
    - Provide the parametrized type for this generic. 문제 해결, generic 구체화를 위한 기본 타입 설명 아래 참고
4. record 적극 도입
- 불변 이슈와 편의성을 위해 기존 클래스에서 record로 전환

From<A,B> = 'from 엔티티'에서 엔티티 부분 , A타입의 객체부터 B타입의 속성을 탐색 (그래서 Root<A,A>이고 Join<A,B>임, Root와 Join 둘다 From 상속)
SingularAttribute<C,D> = 메타모델에서 사용하는 단일 속성 타입 정보 , C = 특정 엔티티, D = C 엔티티의 필드 타입
Path<X> = 메타모델 경로 혹은 루트로부터 탐색된 속성(특정 속성의 경로) , X = 속성 타입, 해당 path가 가르키는 최종 필드 타입

gunwoong1630 committed:
feat: criteriautil 추가

연관관계가 있는 클래스들을 Root에서 연달아 찾아가야할 경우 위와같이 Path를 찾아내는 기능 추가

gunwoong1630 committed:
feat: 최신순 정렬 추가

생성 순으로 휴가 신청 조회기능 추

gunwoong1630 committed:
feat: VacationRequest 상세 조회 및 수정

gunwoong1630 committed:
refactor: 제네릭 누락 및 record 사용 시 배열 사용 부분 리팩토링

1. 제네릭이 누락될경우 컴파일러가 해당 클래스를 raw type으로 두고 타입 추론을 해주지 않습니다. => 제네릭은 항상 명시

2. record는 기본적으로 get,hashCode,equal,toString 메서드가 있는데 원소에 배열이 있을 경우 예상치 못한 결과 발생 가능 => 배열을 List로 수정
